src/realm/query_engine.hpp

```diff
@@ -2101,25 +2141,76 @@ class OrNode : public ParentNode {
     std::vector<std::unique_ptr<ParentNode>> m_conditions;
 
 private:
+    struct ConditionType {
+        ConditionType(const ParentNode& node)
+            : m_col(node.m_condition_column_key.value)
+            , m_type(typeid(node))
+        {
+        }
+        int64_t m_col;
+        std::type_index m_type;
+        bool operator<(const ConditionType& other) const
+        {
+            return this->m_col < other.m_col && this->m_type < other.m_type;
+        }
+        bool operator!=(const ConditionType& other) const
+        {
+            return this->m_col != other.m_col || this->m_type != other.m_type;
+        }
+    };
+
     void combine_conditions(bool ignore_indexes)
     {
+        // Although ColKey is not unique per table, it is not important to consider
+        // the table when sorting here because ParentNode::m_condition_column_key
+        // is only a valid ColKey when the node has a direct condition on a column
+        // of the table this query is running on. Any link query nodes use a special
+        // LinkChain state to store the column path.
         std::sort(m_conditions.begin(), m_conditions.end(), [](auto& a, auto& b) {
-            return a->m_condition_column_key < b->m_condition_column_key;
```

Comment:
The previous sorting code didn't consider the node type in the sort. This might have led to the query not being optimized/combined even if it should have been. The affected queries would have multiple conditions on the same column such as `c == 1 or c == 2 or c == 3 or c > 5`. If the sort put the `c > 5` node between any of the `==` nodes, then not all of the equality nodes would have been combined. The query would have evaluated correctly, just not at peak performance, so I don't think it is worth a changelog note.